grpc/server.go

```diff
@@ -68,3 +68,83 @@ func (s *Server) List(ctx context.Context, req *v1.ListRequest) (*v1.ListRespons
 
 	return repoList.ToProto(), nil
 }
+
+// gRPCChunkSender is a zoekt.Sender that sends small chunks of FileMatches to the provided gRPC stream.
+func gRPCChunkSender(ss v1.WebserverService_StreamSearchServer) zoekt.Sender {
+	f := func(r *zoekt.SearchResult) {
+		result := r.ToProto()
+
+		if len(result.GetFiles()) == 0 { // stats-only result, send it immediately
+			_ = ss.Send(result)
+			return
+		}
+
+		// Otherwise, chunk the file matches into multiple responses
+
+		statsSent := false
+		numFilesSent := 0
+
+		sendFunc := func(filesChunk []*v1.FileMatch) error {
+			numFilesSent += len(filesChunk)
+
+			var stats *v1.Stats
+			if !statsSent { // We only send stats back on the first chunk
+				statsSent = true
+				stats = result.GetStats()
+			}
+
+			progress := result.GetProgress()
+
+			if numFilesSent < len(result.GetFiles()) { // more chunks to come
+				progress = &v1.Progress{
+					Priority: result.GetProgress().GetPriority(),
+
+					// We want the client to consume the entire set of chunks - so we manually
+					// patch the MaxPendingPriority to be >= overall priority.
+					MaxPendingPriority: math.Max(
+						result.GetProgress().GetPriority(),
+						result.GetProgress().GetMaxPendingPriority(),
+					),
+				}
+			}
+
+			// We need to ensure that we send the repository URLs and LineFragments for only the repositories
+			// actually mentioned in this chunk of files.
+			repoURLs := make(map[string]string, len(result.GetRepoUrls()))
+			lineFragments := make(map[string]string, len(result.GetLineFragments()))
+
+			for _, file := range filesChunk {
+				repository := file.GetRepository()
+
+				if url, ok := result.GetRepoUrls()[repository]; ok {
+					repoURLs[repository] = url
+				}
+
+				if fragment, ok := result.GetLineFragments()[repository]; ok {
+					lineFragments[repository] = fragment
+				}
+			}
+
+			response := &v1.SearchResponse{
+				Files:    filesChunk,
+				Stats:    stats,
+				Progress: progress,
+
+				RepoUrls:      repoURLs,
+				LineFragments: lineFragments,
+			}
+
+			return ss.Send(response)
+		}
+
+		chunker := chunk.New(sendFunc)
+		err := chunker.Send(result.GetFiles()...)
+		if err != nil {
+			return
+		}
+
+		_ = chunker.Flush()
```

Comment:
minor: is it worth to introduce a helper to the chunk package which simplifies this "one-shot" sort of use? eg

```suggestion
		_ = chunk.Send(sendFunc, result.GetFiles()...)
```